tests/testthat/test-quosure-join.R

```diff
@@ -0,0 +1,152 @@
+  env1 <- list2env(list(
+    iris1 = iris,
+    iris2 = iris
+  ))
+  env2 <- list2env(list(iris1 = head(iris)))
+
+  q1 <- new_quosure(iris1 <- iris, env = env1)
+  q2 <- new_quosure(iris1 <- head(iris), env = env2)
+
+  testthat::expect_match(check_joinable(q1, q2), "Not possible to join Quosure objects")
+  testthat::expect_error(join(q1, q2), "Not possible to join Quosure objects")
+})
+
+testthat::test_that("join does not duplicate code but adds only extra code", {
+  env <- list2env(list(
+    iris1 = iris,
+    mtcars1 = mtcars
+  ))
+
+  q1 <- new_quosure(code = c(iris1 = "iris1 <- iris", mtcars1 = "mtcars1 <- mtcars"), env = env)
+  q2 <- q1
+  q1 <- eval_code(q1, "iris2 <- iris")
+  q2 <- eval_code(q2, "mtcars2 <- mtcars")
+
+  testthat::expect_true(check_joinable(q1, q2))
+  q <- join(q1, q2)
+
+  testthat::expect_identical(
+    q@code,
+    c(iris1 = "iris1 <- iris", mtcars1 = "mtcars1 <- mtcars", code = "iris2 <- iris", code.1 = "mtcars2 <- mtcars")
+  )
+
+  testthat::expect_equal(
+    as.list(q@env),
+    list(iris1 = iris, iris2 = iris, mtcars1 = mtcars, mtcars2 = mtcars)
+  )
+
+  testthat::expect_identical(q@id, c(q1@id, q2@id[3]))
+})
+
+testthat::test_that("Not possible to join quosures which share some code when one of the shared object was modified", {
+  env <- new.env()
+  env$iris1 <- iris
+  env$mtcars1 <- mtcars
+
+  q1 <- new_quosure(code = c(iris1 = "iris1 <- iris", mtcars1 = "mtcars1 <- mtcars"), env = env)
+  q2 <- q1
+  q1 <- eval_code(q1, "iris2 <- iris")
+  q2 <- eval_code(q2, "mtcars1 <- head(mtcars)")
+
+  testthat::expect_error(join(q1, q2))
+})
+
+testthat::test_that("Quosure objects are mergeable if they don't share any code (identified by id)", {
+  q1 <- new_quosure(code = "a1 <- 1", env = list2env(list(a1 = 1)))
+  q2 <- new_quosure(code = "a1 <- 1", env = list2env(list(a1 = 1)))
+  testthat::expect_true(check_joinable(q1, q2))
+
+  cq <- join(q1, q2)
+  testthat::expect_s4_class(cq, "Quosure")
+  testthat::expect_equal(cq@env, list2env(list(a1 = 1)))
+  testthat::expect_identical(cq@code, c(`initial code` = "a1 <- 1", `initial code.1` = "a1 <- 1"))
+  testthat::expect_identical(cq@id, c(q1@id, q2@id))
+})
+
+testthat::test_that("Quosure objects are mergeable if they share common initial code", {
+  q1 <- new_quosure(code = "a1 <- 1", env = list2env(list(a1 = 1)))
+  q2 <- eval_code(q1, "b1 <- 2")
+  q1 <- eval_code(q1, "a2 <- 3")
+  testthat::expect_true(check_joinable(q1, q2))
+
+  cq <- join(q1, q2)
+  testthat::expect_s4_class(cq, "Quosure")
+  testthat::expect_equal(cq@env, list2env(list(a1 = 1, b1 = 2, a2 = 3)))
+  testthat::expect_identical(
+    cq@code,
+    c(`initial code` = "a1 <- 1", `code` = "a2 <- 3", `code.1` = "b1 <- 2")
+  )
+  testthat::expect_identical(cq@id, union(q1@id, q2@id))
+})
+
+testthat::test_that("Quosure objects aren't mergeable if they share common code proceeded with some other code", {
```

Comment:
So this is not quite true, it's not common code, it's common quosure as here you can join `q5` and `q6` even though they both have common _code_ `c1 <- 1` proceeded by some other code

```
q1 <- new_quosure(code = "a1 <- 1", env = list2env(list(a1 = 1)))
q2 <- new_quosure(code = "b1 <- 1", env = list2env(list(b1 = 1)))

q3 <- new_quosure(code = "c1 <- 1", env = list2env(list(c1 = 1)))
q4 <- new_quosure(code = "c1 <- 1", env = list2env(list(c1 = 1)))


q5 <- join(q1, q3)
q6 <- join(q2, q4)
```

I think we need to be really careful when we mean "code" and when we mean "quosure" throughout the PR